engine/src/main/java/com/google/android/fhir/sync/Sync.kt

```diff
@@ -108,10 +126,89 @@ object Sync {
           }
       }
 
+  /**
+   * Combines the sync state for a periodic sync operation, including work state, progress, and
+   * terminal states.
+   *
+   * @param context The Android application context.
+   * @param workName The name of the periodic sync work.
+   * @param syncJobProgressStateFlow A flow representing the progress of the sync job.
+   * @return A flow of [PeriodicSyncState] combining the sync job states.
+   */
+  @PublishedApi
+  internal fun combineSyncStateForPeriodicSync(
+    context: Context,
+    workName: String,
+    syncJobProgressStateFlow: Flow<SyncJobStatus>,
+    id: UUID,
+  ): Flow<PeriodicSyncState> {
+    val workStateFlow: Flow<WorkInfo.State> = observeWorkState(context, id)
+    val syncJobTerminalStateFlow: Flow<SyncJobStatus> =
+      FhirEngineProvider.getFhirDataStore(context).observeSyncJobTerminalState(workName)
+    val syncJobFlow =
+      combine(syncJobProgressStateFlow, syncJobTerminalStateFlow) {
+        inProgressStatus,
+        terminalStatus,
+        ->
+        if (terminalStatus == SyncJobStatus.Unknown) {
+          inProgressStatus
+        } else {
+          terminalStatus
+        }
+      }
+
+    return combine(workStateFlow, syncJobFlow) { workInfoState, syncJobStatus ->
+      val lastSyncJobStatus =
+        FhirEngineProvider.getFhirDataStore(context).getLastSyncJobStatus(workName)
+      val currentJobState = createCurrentJobState(syncJobStatus, workInfoState)
+      PeriodicSyncState(
+        lastJobState = mapSyncJobStatusToResult(lastSyncJobStatus),
+        currentJobState = currentJobState,
+      )
+    }
+  }
+
+  /**
+   * Combines the sync state for a one-time sync operation, including work state, progress, and
+   * terminal states.
+   *
+   * @param context The Android application context.
+   * @param workName The name of the one-time sync work.
+   * @param syncJobProgressStateFlow A flow representing the progress of the sync job.
+   * @return A flow of [SyncState] combining the sync job states.
+   */
+  @PublishedApi
+  internal fun combineSyncStateForOneTimeSync(
+    context: Context,
+    workName: String,
+    syncJobProgressStateFlow: Flow<SyncJobStatus>,
+    uuid: UUID,
+  ): Flow<SyncState> {
+    val workStateFlow: Flow<WorkInfo.State> = observeWorkState(context, uuid)
+    val syncJobTerminalStateFlow: Flow<SyncJobStatus> =
+      FhirEngineProvider.getFhirDataStore(context).observeSyncJobTerminalState(workName)
+
+    return combine(workStateFlow, syncJobProgressStateFlow, syncJobTerminalStateFlow) {
+      workInfoState,
+      currentSyncJobStatus,
+      terminalSyncJobStatus,
+      ->
+      createSyncState(workInfoState, currentSyncJobStatus, terminalSyncJobStatus)
+    }
+  }
```

Comment:
this function can be simplified a lot now with our WorkInfo.State, SyncJobStatus pair
```suggestion
  internal fun combineSyncStateForOneTimeSync(
    context: Context,
    workName: String,
    workerInfoSyncJobStatusPairFromWorkManagerFlow: Flow<Pair<WorkInfo.State, SyncJobStatus?>>,
  ): Flow<SyncState> {
    val syncJobStatusInDataStoreFlow: Flow<SyncJobStatus?> =
      FhirEngineProvider.getFhirDataStore(context).observeTerminalSyncJobStatus(workName)

    return combine(workerInfoSyncJobStatusPairFromWorkManagerFlow, syncJobStatusInDataStoreFlow) {
      workerInfoSyncJobStatusPairFromWorkManager,
      syncJobStatusFromDataStore,
      ->
      createSyncState(
        workerInfoSyncJobStatusPairFromWorkManager.first,
        workerInfoSyncJobStatusPairFromWorkManager.second,
        syncJobStatusFromDataStore,
      )
    }
  }
```